Userland/Libraries/LibCore/Stream.cpp

```diff
@@ -0,0 +1,538 @@
+    if (seek_result.is_error()) {
+        // Let's try to restore the original position, just in case.
+        auto restore_result = seek(original_position, SeekMode::SetPosition);
+        if (restore_result.is_error()) {
+            dbgln("Core::SeekableStream::size: Couldn't restore initial position, stream might have incorrect position now!");
+        }
+
+        return seek_result.error();
+    }
+
+    TRY(seek(original_position, SeekMode::SetPosition));
+    return seek_result.value();
+}
+
+KResultOr<File> File::open(StringView const& filename, OpenMode mode, mode_t permissions)
+{
+    File file { mode };
+    TRY(file.open_file(filename, permissions));
+    return file;
+}
+
+KResultOr<File> File::adopt_fd(int fd, OpenMode mode)
+{
+    if (fd < 0) {
+        return static_cast<ErrnoCode>(EBADF);
+    }
+
+    if (!has_any_flag(mode, OpenMode::ReadWrite)) {
+        dbgln("Core::File::adopt_fd: Attempting to adopt a file with neither Read nor Write specified in mode");
+        return static_cast<ErrnoCode>(EINVAL);
+    }
+
+    File file { mode };
+    file.m_fd = fd;
+    return file;
+}
+
+KResult File::open_file(StringView const& filename, mode_t permissions)
+{
+    VERIFY(m_fd == -1);
+
+    int flags = 0;
+    if (has_flag(m_mode, OpenMode::ReadWrite)) {
+        flags |= O_RDWR | O_CREAT;
+    } else if (has_flag(m_mode, OpenMode::Read)) {
+        flags |= O_RDONLY;
+    } else if (has_flag(m_mode, OpenMode::Write)) {
+        flags |= O_WRONLY | O_CREAT;
+        bool should_truncate = !has_any_flag(m_mode, OpenMode::Append | OpenMode::MustBeNew);
+        if (should_truncate)
+            flags |= O_TRUNC;
+    }
+
+    if (has_flag(m_mode, OpenMode::Append))
+        flags |= O_APPEND;
+    if (has_flag(m_mode, OpenMode::Truncate))
+        flags |= O_TRUNC;
+    if (has_flag(m_mode, OpenMode::MustBeNew))
+        flags |= O_EXCL;
+    if (!has_flag(m_mode, OpenMode::KeepOnExec))
+        flags |= O_CLOEXEC;
+    if (!has_flag(m_mode, OpenMode::Nonblocking))
+        flags |= O_NONBLOCK;
+
+#ifdef __serenity__
+    int fd = ::serenity_open(filename.characters_without_null_termination(), filename.length(), flags, permissions);
+#else
+    String filename_with_null_terminator { filename };
+    int fd = ::open(filename_with_null_terminator.characters(), flags, permissions);
+#endif
+
+    if (fd < 0) {
+        return static_cast<ErrnoCode>(errno);
+    }
+
+    m_fd = fd;
+    return KSuccess;
+}
+
+bool File::is_readable() const { return has_flag(m_mode, OpenMode::Read); }
+bool File::is_writable() const { return has_flag(m_mode, OpenMode::Write); }
+
+KResultOr<size_t> File::read(Bytes buffer)
+{
+    if (!has_flag(m_mode, OpenMode::Read)) {
+        // NOTE: read(2) manpage says that if the fd is unsuitable for reading,
+        //       the call will return EINVAL. Since we already know whether we
+        //       can or can't read the file, let's avoid a syscall.
```

Comment:
There is no read(2) man page in SerenityOS at this point.
If there were, it would say reading from an fd that's not open for reading will fail with `EBADF`, not `EINVAL`.
Per Dr. POSIX:
> [EBADF] The fildes argument is not a valid file descriptor open for reading.

Reply:
Ah, I completely skipped over the `EBADF` part, my bad. And yeah, by `read(2)` I meant `read(2)` on my host, so I'll correct that to say "POSIX says that".